In [7]:
import pandas as pd
import numpy as np

df = pd.read_csv('../data/tedsd_clean.csv')

print(df.shape)
df.head()

(1474025, 9)


,DISYR,REASON,completed,SUB1,AGE,SEX,RACE,LOS,SERVICES
0,2023,1,1,Alcohol,35-39,Female,Alaska Native,33,7
1,2023,1,1,Alcohol,45-49,Female,Alaska Native,35,7
2,2023,3,0,Alcohol,30-34,Female,Alaska Native,37,7
3,2023,3,0,Alcohol,55-64,Female,Alaska Native,36,7
4,2023,3,0,Alcohol,55-64,Female,Alaska Native,31,7


In [8]:
# Completion rate by primary substance
substance_completion = (
    df.groupby('SUB1')['completed']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'completion_rate', 'count': 'total_episodes'})
    .sort_values('completion_rate', ascending=False)
)

substance_completion['completion_rate'] = substance_completion['completion_rate'].round(3)

print(substance_completion)

                     completion_rate  total_episodes
SUB1                                                
Alcohol                        0.495          436332
Benzodiazepines                0.441           12366
Barbiturates                   0.421            1114
PCP                            0.406            2602
Other Amphetamines             0.382           11260
Hallucinogens                  0.369            1836
Marijuana                      0.369          113384
Cocaine/Crack                  0.343           85622
Methamphetamine                0.340          180249
Inhalants                      0.318             670
Other Tranquilizers            0.304             125
Other Sedatives                0.291            1913
Other Stimulants               0.285            3520
OTC Medications                0.282             379
Other Drugs                    0.253            9237
Heroin                         0.244          223522
Other Opiates                  0.230          

Opioid-related substances (heroin, other opiates, methadone) show the lowest completion rates around 14-24% while alcohol is at the top near 50%. Methamphetamine is a huge concern because of it being the second most-common substance while demonstrating a below average completino rate of 34%.

In [9]:
# Completion rate by gender
gender_completion = (
    df.groupby('SEX')['completed']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'completion_rate', 'count': 'total_episodes'})
)

gender_completion['completion_rate'] = gender_completion['completion_rate'].round(3)
print(gender_completion)

        completion_rate  total_episodes
SEX                                    
Female            0.415          506932
Male              0.432          966184


Males complete at 43.2% and females at 41.5%. This gap seems small, but across nearly 1.5 million records, it presents a big gap. The gap likely reflects unique challenges facing women in treatment rather than differences in motivation or engagement.

In [10]:
# Completion rate by age group
age_order = ['12-14', '15-17', '18-20', '21-24', '25-29', 
             '30-34', '35-39', '40-44', '45-49', '50-54', '55-64', '65+']

age_completion = (
    df.groupby('AGE')['completed']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'completion_rate', 'count': 'total_episodes'})
    .reindex(age_order)
)

age_completion['completion_rate'] = age_completion['completion_rate'].round(3)
print(age_completion)

       completion_rate  total_episodes
AGE                                   
12-14            0.353            6630
15-17            0.362           22332
18-20            0.470           27545
21-24            0.445           77333
25-29            0.422          179628
30-34            0.402          264251
35-39            0.405          241859
40-44            0.418          196192
45-49            0.412          138989
50-54            0.434          121706
55-64            0.482          165609
65+              0.562           31951


The pattern is U-shaped. Adolescents (12-14) have the lowest completion rate at 35%. Middle aged patients cluster around 40-42%. Older adults (65+) have the highest rate at 56%. Older patients likely benefit from stronger self-motivation and more stable life circumstances.

In [11]:
# Average length of stay by completion status
los_analysis = (
    df.groupby('completed')['LOS']
    .agg(['mean', 'median', 'count'])
    .round(1)
)
los_analysis.index = ['Not Completed', 'Completed']
print(los_analysis)

               mean  median   count
Not Completed  18.0    17.0  846128
Completed      21.1    28.0  627897


Completers have a median stay of 28 days versus 17 days for non-completers. The median is more meaningful than the mean here because a few very long stays skew the average. The 11-day gap suggests that early dropout in the first two weeks is the biggest driver of non-completion.

In [13]:
# df is already loaded, create df_clean reference
df_clean = df.copy()

# Export analysis results for Power BI
substance_completion.to_csv('../data/substance_completion.csv')
gender_completion.to_csv('../data/gender_completion.csv')
age_completion.to_csv('../data/age_completion.csv')
los_analysis.to_csv('../data/los_analysis.csv')
df_clean.to_csv('../data/tedsd_clean.csv', index=False)

print("All files exported successfully")

All files exported successfully
